go/pools/smartconnpool/pool.go

```diff
@@ -0,0 +1,709 @@
+	// wait for one of them
+	if conn == nil && pool.borrowed.Load() > 0 {
+		start := time.Now()
+		conn, err = pool.wait.waitForConn(ctx, setting)
+		if err != nil {
+			if conn != nil {
+				// our context expired but we managed to get a conn from
+				// the waitlist; put it back!
+				pool.put(conn)
+			}
+			return nil, ErrTimeout
+		}
+		pool.recordWait(start)
+	}
+	// no connections available and no connections to wait for (pool is closed)
+	if conn == nil {
+		return nil, ErrTimeout
+	}
+
+	// ensure that the setting applied to the connection matches the one we want
+	connSetting := conn.Conn.Setting()
+	if connSetting != setting {
+		// if there's another setting applied, reset it before applying our setting
+		if connSetting != nil {
+			pool.Metrics.diffSetting.Add(1)
+
+			err = conn.Conn.ResetSetting(ctx)
+			if err != nil {
+				conn.Close()
+				err = pool.connReopen(ctx, conn, time.Now())
+				if err != nil {
+					pool.active.Add(-1)
+					return nil, err
+				}
+			}
+		}
+		// apply our setting now; if we can't we assume that the conn is broken
+		// and close it without returning to the pool
+		if err := conn.Conn.ApplySetting(ctx, setting); err != nil {
+			conn.Close()
+			pool.active.Add(-1)
+			return nil, err
+		}
+	}
+
+	pool.borrowed.Add(1)
+	return conn, nil
+}
+
+// SetCapacity changes the capacity (number of open connections) on the pool.
+// If the capacity is smaller than the number of connections that there are
+// currently open, we'll close enough connections before returning, even if
+// that means waiting for clients to return connections to the pool.
+func (pool *ConnPool[C]) SetCapacity(newcap int64) {
+	if newcap < 0 {
+		panic("negative capacity")
+	}
+
+	var oldcap int64
+
+	for {
+		oldcap = pool.capacity.Load()
+		if oldcap == newcap {
+			return
+		}
+		if pool.capacity.CompareAndSwap(oldcap, newcap) {
+			break
+		}
+	}
+
+	// close connections until we're under capacity
+	for pool.active.Load() > newcap {
+		// try closing from connections which are currently idle in the stacks
+		conn := pool.getFromSettingsStack(nil)
+		if conn == nil {
+			conn, _ = pool.clean.Pop()
+		}
+		// if we can't find connections in the stacks, try waiting for clients
+		// to return one
+		if conn == nil {
+			conn, _ = pool.wait.waitForConn(context.Background(), nil)
+		}
+		if conn == nil {
+			continue
+		}
+		conn.Close()
+		pool.active.Add(-1)
+	}
+}
+
+func (pool *ConnPool[C]) closeIdleResources(now time.Time) {
+	timeout := pool.IdleTimeout()
+	if timeout == 0 {
+		return
+	}
+	if pool.Capacity() == 0 {
+		return
+	}
+
+	var conns []*Pooled[C]
+
+	closeInStack := func(s *connStack[C]) {
+		conns = s.PopAll(conns[:0])
+		slices.Reverse(conns)
+
+		for _, conn := range conns {
+			if conn.timeUsed.Add(timeout).Sub(now) < 0 {
+				pool.Metrics.idleClosed.Add(1)
+
+				conn.Close()
+				if err := pool.connReopen(context.Background(), conn, now); err != nil {
+					pool.active.Add(-1)
+					continue
+				}
```

Comment:
Why are we reopening a connection when it is closed because it was idle too long? Is this so that we have a connection open the next time a user requests one?
Shouldn't we just reduce the number of open connections if the connection is not being used? If the load of the entire application went down, then we don't need to reopen connections that have been closed due to idleness.

Reply:
Good catch! I copied this functionality from the old resource pool, where it makes a bit more sense because connections are continuously cycled, so re-connecting the idle one is more useful because it'll get (eventually) pulled from the pool even if the system is not at capacity.

With the new LIFO design, re-opening the connection here will probably only cause it to be closed again in the future. I've changed the code so it doesn't re-open. :ok_hand: